File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    # 각 사람별 인덱스 값
    f_idx= {v:i for i, v in enumerate(friends)}
    # 받는 선물 개수 배열
    answer = [0] * len(friends)
    # 각 사람별 선물 지수 배열
    n_dict = [0] * len(friends)
    # 각 사람별로 다른 사람들에게 준 선물 개수 나타낸 배열
    gt_dict = [[0] * len(friends) for i in range(len(friends))]
    
    # 주고받은 선물 개수 업데이트
    for gift in gifts:
        give, take = gift.split()
        gt_dict[f_idx[give]][f_idx[take]] += 1

    # 선물 지수 계산 및 업데이트
    for i in range(len(friends)):
        # i의 선물 지수 = i가 준 선물 개수 - i가 받은 선물 개수
        n_dict[i] = sum(gt_dict[i]) - sum([k[i] for k in gt_dict])

    # 선물 비교를 통해 받는 선물 업데이트
    for i in range(len(friends)):
        for j in range(len(friends)):
            # 준 것 > 받은 것이면 다음달에 +1
            if gt_dict[i][j] > gt_dict[j][i]: answer[i] += 1
            # 주고 받은 것이 같지만 선물 지수가 더 높으면 다음달에 +1
            elif gt_dict[i][j] == gt_dict[j][i]:
                if n_dict[i] > n_dict[j]: answer[i] += 1

    return max(answer)

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.js
function solution(friends, gifts) {
    let max = 0;
    // 선물 준 사람을 key, 받은 사람 배열을 value로 하는 객체
    let gt_dict = new Map();
    // 각 사람 별 선물 지수 객체
    let n_dict = new Map();

    friends.forEach(v => {
        gt_dict.set(v, []);
        n_dict.set(v, 0);
    })

    gifts.forEach(gift => {
        let [give, take] = gift.split(" ")
        // 선물 받은 사람들 value로 배열에 넣기
        gt_dict.get(give).push(take);
        // 선물 준 사람들은 +1
        n_dict.set(give, n_dict.get(give)+1);
        // 선물 받은 사람들은 -1
        n_dict.set(take, n_dict.get(take)-1);
    })
    
    for(let i = 0; i < friends.length; i++){
        let count = 0;
        for(let j = 0; j < friends.length; j++){
            // i가 j에게 준 갯수
            let give = gt_dict.get(friends[i]).filter((v) => v == friends[j]).length;
            // i가 j에게 받은 갯수
            let take = gt_dict.get(friends[j]).filter((v) => v == friends[i]).length;
            // 준 것 > 받은 것이면 i의 count++
            if(give > take) count++;
            // 주고 받은 수가 같지만 i의 선물 지수가 더 크면 count++
            else if(give == take){
                if(n_dict.get(friends[i]) > n_dict.get(friends[j])) count++;
            }
        }
        // 가장 많이 받은 선물 수 업데이트
        if(count > max) max = count;
    }

    return max;
}